### PR TITLE
Add Sapho Chart

### PR DESCRIPTION
**Pull request by lodotek**

### CODE REVIEW

**Comment by prydonius** [comment]
@k8s-bot ok to test

**Comment by prydonius** [comment]
Thanks for the submission @lodotek!

Can you remove the vendored mysql in the `charts/` directory? You should be able to use `helm dep update` to pull in the packaged mysql chart which will be ignored in git. The CI will automatically do this when packaging the chart for the repository.

Also, AFAICT this meets the [requirements](https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements) to be included in the stable repo, so you can move this to the stable folder :).

File: incubator/sapho/values.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  limits:
+    cpu: 300m
+    memory: 1010Mi
+  requests:
+    cpu: 300m
+    memory: 1010Mi
+saphoDBport: 3306
+saphoDBtype: mysql
+saphoDBuser: root
+saphoDBpass: 7iro7Ux6Xl
```
Comment (prydonius) [inline incubator/sapho/values.yaml]: looks like this isn't being used

**Comment by lodotek** [comment]
@prydonius I have made the requested changes, please test again. 

Thank you!

**Comment by aronchick** [comment]
@prydonius I'd love to demo this soon - any chance this is ready to go?

File: incubator/sapho/values.yaml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  type: LoadBalancer
+  externalPort: 80
+  internalPort: 8080
+
+resources:
+  limits:
+    cpu: 300m
+    memory: 1010Mi
+  requests:
+    cpu: 300m
```
Comment (viglesiasce) [inline incubator/sapho/values.yaml]: We should bump this up by default. The initial startup is taking minutes and would likely benefit from more gas.

**Review by viglesiasce — CHANGES_REQUESTED** [review]
The sapho pod never went to a healthy state. 

The following was found in the log:

```console
Cannot start Sapho:
- Sapho context cannot be started. Missing injection context.
Error creating bean with name 'entityManagerFactory' defined in class path resource [spring-sapho-database.xml]: Cannot resolve reference to bean 'dataSource' while setting bean property 'dataSource'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'dataSource' defined in class path resource [spring-sapho-database.xml]: Bean instantiation via constructor failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.zaxxer.hikari.HikariDataSource]: Constructor threw exception; nested exception is com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: Communications link failure
```

File: incubator/sapho/values.yaml (hunk 2 above)
Comment (lodotek) [reply incubator/sapho/values.yaml]: I have upped the values a bit, let's see if that helps..

**Review by lodotek — COMMENTED** [review]
For some reason the mysql container failed to bind the PVC on every first attempt it appeared, resulting in our sapho container from trying to prematurely connect to MySQL. As a temporary solution, I have disable the persistent volume support for MySQL. 

Also tweaked the resource settings...

**Comment by lodotek** [comment]
Anything else needed from me to get this PR approved?

**Comment by prydonius** [comment]
@lodotek MySQL was failing for me due to `ERROR 1396 (HY000) at line 1: Operation CREATE USER failed for 'root'@'%'` when persistence is enabled. Both MySQL and Sapho started up fine for me when I removed the `mysqlUser: root` setting from `values.yaml` and re-enabled persistence.

If you're having issues with PVCs getting bound, does your cluster support dynamic provisioning?

File: incubator/sapho/templates/NOTES.txt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+     NOTE: It may take a few minutes for the LoadBalancer IP to be available.
+           You can watch the status of by running 'kubectl get svc -w {{ template "fullname" . }}'
+  export SERVICE_IP=$(kubectl get svc --namespace {{ .Release.Namespace }} {{ template "fullname" . }} -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
+  echo http://$SERVICE_IP:{{ .Values.service.externalPort }}
+{{- else if contains "ClusterIP"  .Values.service.type }}
+  export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app={{ template "fullname" . }}" -o jsonpath="{.items[0].metadata.name}")
+  echo "Visit http://127.0.0.1:8080 to use your application"
+  kubectl port-forward $POD_NAME 8080:{{ .Values.service.externalPort }}
+{{- end }}
+MySQL can be accessed via port 3306 on the following DNS name from within your cluster:
```
Comment (prydonius) [inline incubator/sapho/templates/NOTES.txt]: I'm not sure you really need to mention how MySQL can be accessed here. If you do though, can you add a line break, as it looks like it's part of the previous statement:

```
  export SERVICE_IP=$(kubectl get svc --namespace default sapho-sapho -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
  echo http://$SERVICE_IP:80
MySQL can be accessed via port 3306 on the following DNS name from within your cluster:
sapho-sapho.default.svc.cluster.local
```

File: incubator/sapho/values.yaml [hunk 4]
```diff
@@ -0,0 +1,34 @@
+# Default values for sapho.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+replicaCount: 1
+image: sapho/ops-docker-tomcat:8-jre8
+
+mysql:
+  mysqlUser: root
```
Comment (prydonius) [inline incubator/sapho/values.yaml]: This setting causes MySQL to crash, `root` is the default user created so with this it tries to create the root user twice. It should start up fine without this set.

File: incubator/sapho/templates/deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+          value: {{ .Values.saphoDBuser }}
+        - name: SAPHO_DB_PASSWORD
+          valueFrom:
+            secretKeyRef:
+              name: {{ template "mysql.fullname" . }}
+              key: mysql-root-password
+        ports:
+        - name: http
+          containerPort: 8080
+#        livenessProbe:
```
Comment (prydonius) [inline incubator/sapho/templates/deployment.yaml]: @lodotek this can be uncommented again right?

**Review by lodotek — COMMENTED** [review]
incubator/sapho/values.yaml:
Removed line 8 specifying mysqlUser
Changed Line 9, re-enabled persistence...

incubator/sapho/templates/NOTES.txt:
Removed section about MySQL

File: incubator/sapho/templates/deployment.yaml (hunk 5 above)
Comment (lodotek) [reply incubator/sapho/templates/deployment.yaml]: Yes, I think you are right. I will test and push the changes.Thanks for all the help, sorry I am kind of a noob haha...

File: incubator/sapho/Chart.yaml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
+description: A Helm chart for Kubernetes
```
Comment (prydonius) [inline incubator/sapho/Chart.yaml]: Could you change this to a short description of Sapho?

File: incubator/sapho/README.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
+# Sapho
+
+[Sapho](https://www.sapho.com/) is awesome.
```
Comment (prydonius) [inline incubator/sapho/README.md]: Would be good to have a better description of what Sapho is here.

File: incubator/sapho/README.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+This chart bootstraps a [Sapho](https://bitbucket.org/sapho/ops-docker-tomcat/) deployment on a [Kubernetes](http://kubernetes.io) cluster using the [Helm](https://helm.sh) package manager.
+
+It also packages the [MySQL chart](https://github.com/kubernetes/charts/tree/master/stable/mysql) which is required for bootstrapping a MySQL deployment for the database requirements of the Sapho application.
+
+## Prerequisites
+
+- Kubernetes 1.4+ with Beta APIs enabled
+- PV provisioner support in the underlying infrastructure
+
+## Get this chart
```
Comment (prydonius) [inline incubator/sapho/README.md]: Can you update these sections to use the Helm repo? For example, see: https://github.com/kubernetes/charts/tree/master/stable/mariadb#installing-the-chart

**Comment by prydonius** [comment]
@lodotek is there any reason you want to keep this in incubator? I would move it to stable since it meets the requirements.

**Comment by lodotek** [comment]
@prydonius Thank you for all you help and suggestions.  Regarding your comment about why am I keeping this in incubator. I thought this PR needs to be approved first, no?

File: incubator/sapho/README.md (hunk 7 above)
Comment (lodotek) [reply incubator/sapho/README.md]: Thank you for the reminder, this has been completed! :-)

**Comment by prydonius** [comment]
@lodotek no worries! As long as a chart meets the requirements listed in [CONTRIBUTING.md](https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#how-to-contribute-a-chart) it can be submitted to the stable repo. The incubator repo is for charts that don't meet these requirements (for example, using k8s alpha features), but can be improved on until they are ready to be moved into stable.

So since this already meets the requirements, it can be submitted directly to stable. Sorry for the confusion there.

**Comment by prydonius** [comment]
@viglesiasce this worked for me, can you check again if you're seeing the same issue you had before?

**Comment by viglesiasce** [comment]
Im still seeing the same issue as before when installing in a fresh cluster. This is likely due to the MySQL image being pulled for the first time and taking a while to come up. 

```console
Caused by: java.net.SocketTimeoutException: connect timed out
        at java.net.PlainSocketImpl.socketConnect(Native Method) ~[na:1.8.0_111]
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350) ~[na:1.8.0_111]
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206) ~[na:1.8.0_111]
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188) ~[na:1.8.0_111]
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[na:1.8.0_111]
        at java.net.Socket.connect(Socket.java:589) ~[na:1.8.0_111]
        at com.mysql.jdbc.StandardSocketFactory.connect(StandardSocketFactory.java:211) ~[mysql-connector-java-5.1.39.jar:5.1.39]
        at com.mysql.jdbc.MysqlIO.<init>(MysqlIO.java:300) ~[mysql-connector-java-5.1.39.jar:5.1.39]
        ... 63 common frames omitted
00:11:59.735 [localhost-startStop-1] ERROR s.s.SaphoServletContextListener -

********************************************************************************

Cannot start Sapho:
- Sapho context cannot be started. Missing injection context.
Error creating bean with name 'entityManagerFactory' defined in class path resource [spring-sapho-database.xml]: Cannot resolve reference to bean 'dataSource' while setting bean property 'dataSource'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'dataSource' defined in class path resource [spring-sapho-database.xml]: Bean instantiation via constructor failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.zaxxer.hikari.HikariDataSource]: Constructor threw exception; nested exception is com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: Communications link failure

The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.
```

Eventually the pod was restarted due to failing its liveness check. At that point it came up fine.

@lodotek can you add something to the NOTES.txt that instructs user to be patient as the app and db come up?

**Comment by viglesiasce** [comment]
@k8s-bot e2e test this

**Review by viglesiasce — APPROVED** [review]

**Pull request merged**